streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -484,6 +539,7 @@ private void initTopology() {
     public void suspend() {
         log.debug("Suspending");
         suspend(true);
+        transitionTo(State.SUSPENDED);
```

Comment:
Hmm, could task state be CREATED or SUSPENDED as well when suspending? I think SUSPENDED is possible since the onPartitionsRevoked maybe called multiple times before onAssignment during unsuccessful rebalances.

Also depending on the current state the private suspending impl should be slightly different:

1. if it is in RUNNING, just execute the current steps.
2. if it is in RESTORING, we do not need to close topology since it is not initialized yet, only committing / checkpoint is sufficient.
3. if it is in SUSPEND, we do not need to do anything.

Reply:
Good feedback. Right now `closeTopology` is actually a no-op if the topology hasn't been initialized, so there's no need to actually check for case 2 (was restoring or created).

I've added the early-return for if we're already suspended.

While looking into this, I noticed something unsafe about the `close()` methods. If we start closing (clean) and get some exception in the middle of the close, we had better never use the task again, since it's been half-closed. I added a "CLOSING" state for us to transition to as soon as we enter `close()`, and once you're in "CLOSING", the only valid next step is "CLOSED". Hopefully, this helps provide a little safety. Not pressing, but maybe we should do the same for "SUSPEND{ING/ED}".

Reply:
That sounds good to me; I'm wondering if we want to transit to this state if e.g. `commit` throws and before the thread close it as dirty? If yes maybe we can rename it to `ERROR`.